Rules/CommonScripts/Spectator.as

```diff
@@ -15,101 +27,188 @@ CPlayer@ targetPlayer()
+
 			if (zoomTarget > 1.0f)
+			{
 				zoomTarget = 1.0f;
+			}
 			else
+			{
 				zoomTarget = 0.5f;
-
+			}
 		}
-
 	}
 	else
 	{
 		timeToScroll -= getRenderApproximateCorrectionFactor();
 	}
 
-	Vec2f pos = camera.getPosition();
-
-	if (Maths::Abs(camera.targetDistance - zoomTarget) > 0.001f)
-	{
-		camera.targetDistance = (camera.targetDistance * (3 - getRenderApproximateCorrectionFactor() + 1.0f) + (zoomTarget * getRenderApproximateCorrectionFactor())) / 4;
-	}
-	else
-	{
-		camera.targetDistance = zoomTarget;
-	}
-
-	f32 camSpeed = getRenderApproximateCorrectionFactor() * 15.0f / zoomTarget;
-
-	//Move the camera using the action movement keys
+	//move camera using action movement keys
 	if (controls.ActionKeyPressed(AK_MOVE_LEFT))
 	{
 		pos.x -= camSpeed;
 		SetTargetPlayer(null);
+		timeToCinematic = CINEMATIC_TIME;
 	}
 	if (controls.ActionKeyPressed(AK_MOVE_RIGHT))
 	{
 		pos.x += camSpeed;
 		SetTargetPlayer(null);
+		timeToCinematic = CINEMATIC_TIME;
 	}
 	if (controls.ActionKeyPressed(AK_MOVE_UP))
 	{
 		pos.y -= camSpeed;
 		SetTargetPlayer(null);
+		timeToCinematic = CINEMATIC_TIME;
 	}
 	if (controls.ActionKeyPressed(AK_MOVE_DOWN))
 	{
 		pos.y += camSpeed;
 		SetTargetPlayer(null);
+		timeToCinematic = CINEMATIC_TIME;
 	}
 
-    if(controls.isKeyJustReleased(KEY_LBUTTON))
+    if (controls.isKeyJustReleased(KEY_LBUTTON))
     {
         waitForRelease = false;
-
     }
 
-	//Click on players to track them or set camera to mousePos
+	if (timeToCinematic > 0) //player-controlled zoom
+	{
+		if (Maths::Abs(camera.targetDistance - zoomTarget) > 0.001f)
+		{
+			camera.targetDistance = (camera.targetDistance * (3.0f - getRenderApproximateCorrectionFactor() + 1.0f) + (zoomTarget * getRenderApproximateCorrectionFactor())) / 4.0f;
+		}
+		else
+		{
+			camera.targetDistance = zoomTarget;
+		}
+
+		if (AUTO_CINEMATIC)
+		{
+			timeToCinematic -= getRenderApproximateCorrectionFactor();
+		}
+	}
+	else //cinematic camera
+	{
+		//by default, view entire map from center
+		posTarget = Vec2f(mapDim.x, mapDim.y) / 2.0f;
+		f32 zoomW = calculateZoomLevelW(mapDim.x);
+		f32 zoomH = calculateZoomLevelH(mapDim.y);
+		zoomTarget = Maths::Min(zoomW, zoomH);
+		zoomTarget = Maths::Clamp(zoomTarget, 0.5f, 2.0f);
+
+		CBlob@[] blobs;
+		getBlobs(@blobs);
+		calculateImportance(blobs);
+		blobs = sortBlobsByImportance(blobs);
+
+		CBlob@[] players;
+		getBlobsByTag("player", @players);
+
+		if (
+			this.isMatchRunning() && !this.isWarmup() && !this.isGameOver() && //game running
+			!focusOnBlob(blobs) && //not focusing on a blob
+			players.length > 0 //players exist
+		) {
+			u32 maxDistX = 0.0f;
+			u32 maxDistY = 0.0f;
+
+			//calculate mean position of all players
+			posTarget = Vec2f(0, 0);
+			for (uint i = 0; i < players.length; i++)
+			{
+				CBlob@ blob = players[i];
+
+				posTarget += blob.getInterpolatedPosition();
+
+				//look for largest distance between two players
+				for (uint j = 0; j < players.length; j++)
```

Comment:
start at i+1

Reply:
and the end condition above that should be player.lenght-1

Reply:
I can't do `player.length - 1` because the same loop is used for calculating the average position of all players. This would skip over the last player and make the camera sit off centre.

Reply:
It's not an important enough improvement to worry too much about it just leave` i < player.length`. You can still do j = i+1. (If you really wanted you can do `posTarget += players.last().getInterpolatedPosition()` before calculating the average.